File: WikiCodeParser/Elements/MdHeadingElement.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using WikiCodeParser.Nodes;

namespace WikiCodeParser.Elements
{
    public class MdHeadingElement : Element
    {
        public override bool Matches(Lines lines)
        {
            var value = lines.Value();
            return value.Length > 0 && value[0] == '=';
        }

        public override INode Consume(Parser parser, ParseData data, Lines lines, string scope)
        {
            var value = lines.Value().Trim();
            var res = Regex.Match(value, @"^(=+)(.*?)=*$", RegexOptions.IgnoreCase);
            var level = Math.Min(6, res.Groups[1].Value.Length);
            var text = res.Groups[2].Value.Trim();

            var id = GenerateUniqueHeaderID(data, text);
            return new HeadingNode(level, id, text);
        }

        private static string GenerateUniqueHeaderID(ParseData data, string text)
        {
            text = text.ToLower().Replace(' ', '-');
            text = Regex.Replace(text, @"[^0-9a-z\-]", "");
            text = Regex.Replace(text, @"\-{2,}", "-");
            if (text.Length < 3) text = "h" + Guid.NewGuid().ToString("N").Substring(6, 12).ToLower();
            if (Char.IsDigit(text[0])) text = "_" + text;

            const string key = nameof(MdHeadingElement) + ".IdList";
            var list = data.Get(key, () => new List<string>());
            var id = text;
            var num = 2;
            while (list.Contains(id)) id = text + "-" + num++;
            list.Add(id);
            return id;
        }

        public class HeadingNode : INode
        {
            public int Level { get; }
            public string ID { get; }
            public string Text { get; }

            public HeadingNode(int level, string id, string text)
            {
                Level = level;
                ID = id;
                Text = text;
            }

            public string ToHtml()
            {
                return $"<h{Level} id=\"{ID}\">"
                       + System.Web.HttpUtility.HtmlEncode(Text)
                       + $"</h{Level}>";
            }

            public string ToPlainText()
            {
                return Text + "\n" + new string('-', Text.Length);
            }

            public IEnumerable<INode> GetChildren()
            {
                yield break;
            }
        }
    }
}

File: javascript/src/Elements/MdHeadingElement.ts
import { Parser } from '..';
import { Lines } from '../Lines';
import { INode } from '../Nodes/INode';
import { ParseData } from '../ParseData';
import { TagParseContext } from '../TagParseContext';
import { Element } from './Element';

class HeadingNode implements INode {
    public Level : number;
    public ID : string;
    public Text : INode;
    constructor(level : number, id : string, text : INode) {
        this.Level = level;
        this.ID = id;
        this.Text = text;
    }

    ToHtml(): string {
        return `<h${this.Level} id="${this.ID}">${this.Text.ToHtml()}</h${this.Level}>`;
    }

    ToPlainText(): string {
        const plain = this.Text.ToPlainText().replace(/\n/g, ' ');
        return plain + '\n' + '-'.repeat(plain.length);
    }

    GetChildren(): INode[] {
        return [ this.Text ];
    }

    ReplaceChild(i: number, node: INode): void {
        if (i != 0) throw new Error('Argument out of range');
        this.Text = node;
    }

    HasContent(): boolean {
        return true;
    }
}

export class MdHeadingElement extends Element {
    public Matches(lines: Lines): boolean {
        const value = lines.Value();
        return value.length > 0 && value.startsWith('=');
    }
    public Consume(parser: Parser, data: ParseData, lines: Lines, scope: string): INode {
        const value = lines.Value().trim();
        const res = /^(=+)(.*?)=*$/i.exec(value);
        const level = Math.min(6, res[1].length);
        const text = res[2].trim();

        let contents = parser.ParseTags(data, text, scope, TagParseContext.Inline);
        contents = parser.RunProcessors(contents, data, scope);
        const id = MdHeadingElement.GetUniqueAnchor(data, contents.ToPlainText());
        return new HeadingNode(level, id, contents);
        
    }

    private static GetUniqueAnchor(data : ParseData, text : string) : string
    {
        const key = MdHeadingElement.name + '.IdList';
        const anchors = data.Get(key, () => new Set<string>());

        const id = text.replace(/[^\da-z?/:@\-._~!$&'()*+,;=]/ig, '_');
        let anchor = id;
        let inc = 1;
        do {
            // Increment if we have a duplicate
            if (!anchors.has(anchor)) break;
            inc++;
            anchor = `${id}_${inc}`;
        // eslint-disable-next-line no-constant-condition
        } while (true);

        anchors.add(anchor);
        return anchor;
    }
}